### DISCUSSION TITLE
ESPNow async support: awrite concurrency

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#12132] [ESP32]
The docs are unclear whether the following is legitimate without a `Lock` to protect the `asend` calls:
```python
async def task1():
    # code
    await espnow.asend(mac, msg)

async def task2():
    # code
    await espnow.asend(mac, msg)
```
If it were a UART you'd definitely need a `Lock`.
On my reading of the `aioespnow` code the writes are atomic and a `Lock` is not needed.

Whether or not I'm correct, the docs could do with clarifying.

**Comment by bidrohini** [comment]
I think you can post this too the esp forum too: https://www.esp32.com/search.php?search_id=unanswered

**Comment by glenn20** [comment]
Yes, this should not need a `Lock` (unless I have misunderstood something fundamental about `uasyncio`). As noted in the code, it does rely on the internal implementation in micropython/extmod/asyncio/core.py:IOQueue.queue_write().

**Comment by bulletmark** [comment]
@peterhinch, are you asking for the case where the`mac` in `task1()` and `task2()` is the same address?

**Reply by peterhinch** [reply]
I don't believe a lock is required regardless of `mac`. The `asend` method in `aioespnow` calls `send` which is a blocking function, so there is no concurrency. The design of `asend` ensures that `send` cannot be called when a handshake is in progress. (This is on my reading of the code).

**Reply by bulletmark** [reply]
It wasn't clear whether your question was asking about concurrent messages to the same address, or concurrent messages to different addresses because I guess there could potentially be different answers. Although I agree that it probably works fine for both cases.
